napari/_qt/threading.py

```diff
@@ -0,0 +1,563 @@
+                raise TypeError(
+                    f'"connections[{key!r}]" is not a callable function'
+                )
+            getattr(worker, key).connect(val)
+
+    def start():
+        start_worker(worker)
+
+    if start_thread:
+        start()
+    else:
+        worker.start = start
+    return worker
+
+
+def thread_worker(
+    function: Optional[Callable] = None,
+    worker_class: Type[WorkerBase] = Worker,
+) -> Callable:
+    """Decorator that runs a function in a seperate thread when called.
+
+    When called, the decorated function returns a :class:`Worker`.  See
+    :func:`worker_factory` for additional keyword arguments that can be used
+    when calling the function.
+
+    The returned worker will have these signals:
+
+        - *started*: emitted when the work is started
+        - *finished*: emitted when the work is finished
+        - *yielded*: emitted with yielded values (if generator used)
+        - *returned*: emitted with return value
+        - *errored*: emitted with error object on Exception
+        - *paused*: emitted when a running job has successfully paused
+        - *resumed*: emitted when a paused job has successfully resumed
+        - *aborted*: emitted when a running job is successfully aborted
+
+    If the decorated function is a generator, the (default) returned worker
+    will also provide these methods:
+
+        - *quit*: ask the thread to quit
+        - *toggle_paused*: toggle the running state of the thread.
+        - *send*: send a value into the generator.  (This requires that your
+          decorator function uses the ``value = yield`` syntax)
+
+        # optionally
+        - *start*: if you call your function with ``func(start_thread=False)``,
+          it will also have a `.start()` method that can be used to start
+          execution of the function in another thread.  (useful if you need
+          to connect callbacks to signals prior to execution)
+
+    Parameters
+    ----------
+    func : callable
+        Function to call in another thread.  For communication between threads
+        may be a generator function.
+
+    Returns
+    -------
+    callable
+        function that creates a worker, puts it in a new thread and returns
+        the worker instance.
+
+    Example
+    -------
+
+    .. code-block:: python
+
+        @thread_worker
+        def long_function(start, end):
+            # do setup here.
+            yield # Critical to yield before computation when using generator
+
+            # do work, periodically yielding
+            i = start
+            while i <= end:
+                time.sleep(0.1)
+                yield i
+
+            # do teardown
+            return 'anything'
+
+        # call the function to start running in another thread.
+        worker = long_function(start_thread=False)
+        # connect signals if desired
+        worker.start()
+    """
+
+    def inner_func(func: Callable) -> Callable:
```

Comment:
`inner_func` is certainly not descriptive enough... In some sense, this is the decorator, right? Also, why is `function` optional? It feels like one level of nesting too many...

Reply:
fixed with the lovely `toolz.curry`